Fix typed array export

KoBeWi opened the pull request:
*Borrowed* from #67454
Fixes #62916

The issue has multiple duplicates, so I think it's better to have it fixed sooner than later. I rebased the PR and applied my comments. I also removed references to `_editor_prop_ptr_indices`, because it doesn't seem to be used anywhere (leftover from some old code?).

The code overall needs checking probably, I mostly copied the PR and made it work (it just needed a fix in array assigning).

KoBeWi committed:
Fix typed array export

Co-authored-by: Guilherme Sousa <guilherme.sousa1994@gmail.com>

YuriSizov commented:
> The issue has multiple duplicates, so I think it's better to have it fixed sooner than later.

Yes, but for all intents and purposes, this is a missing feature, not a bug. So it won't be too bad if we address this in 4.1.

> The code overall needs checking probably, I mostly copied the PR and made it work (it just needed a fix in array assigning).

Reduz gave it a cursory look before, and said

> that looks pretty good, but I think I would fix this after 4.0, at this point I think its too risky to touch PackedScene
> any issue and you can get corrupted files

KoBeWi commented:
I wouldn't call it a missing feature. We do support node export and array export, they are just broken when used together.
But yeah, the changes seem a bit risky.

GuilhermeGSousa commented:
Thanks for doing this, I didn't have time to work on my PR but I'll be testing you branch today!

akien-mga commented on scene/resources/packed_scene.cpp:
```diff
@@ -238,16 +238,38 @@ Node *SceneState::instantiate(GenEditState p_edit_state) const {
-							node->set(META_POINTER_PROPERTY_BASE + String(snames[name_idx]), props[nprops[j].value]);
+
+						if (prop_variant.get_type() == Variant::ARRAY) {
+							if (Engine::get_singleton()->is_editor_hint()) {
+								// If editor, just set the metadata and be it
+								node->set(prop_name, prop_variant);
+								continue;
+							}
```
Why does it need to be special cased here? The only difference seems to be not using the `META_POINTER_PROPERTY_BASE` prefix, does it break Node arrays?

KoBeWi replied:
This will assign the raw array of NodePaths. `META_POINTER_PROPERTY_BASE` doesn't seem to work here. The condition can't be moved elsewhere and removing it crashes, so it's likely the only way.

akien-mga commented on scene/resources/packed_scene.cpp:
```diff
@@ -416,7 +438,21 @@ Node *SceneState::instantiate(GenEditState p_edit_state) const {
 		Node *other = dnp.base->get_node_or_null(dnp.path);
-		dnp.base->set(dnp.property, other);
+		const String string_property = dnp.property;
+		if (string_property.contains("/indices/")) {
+			const Vector<String> properties = string_property.split("/");
+			Array array = dnp.base->get(properties[0]);
+
+			if (array.size() >= properties[2].to_int()) {
+				array.push_back(other);
+			} else {
+				array.set(properties[2].to_int(), other);
+			}
```
Is it possible for a non-Array property path to include `/indices/` too, and possibly trigger this logic in an unexpected way and crash?

Or can an Array property include `/` in the first place, leading to the `"/indices/`" + itos(k)` suffix being at different indices when splitting by `/`?

KoBeWi replied:
> Is it possible for a non-Array property path to include /indices/ too, and possibly trigger this logic in an unexpected way and crash?

Yes and no. This code only runs for "deferred NodePath properties", i.e. properties stored as NodePath that are assigned a node after the scene is complete. This limits severely what properties are affected by this code. Users can't use `/` in their properties without using `_get_property_list()`, so it's extremely unlikely that this logic is triggered when it shouldn't. And even if it does, assigning non-Array Variant will result in empty array, I think. So it shouldn't crash.

> Or can an Array property include / in the first place, leading to the "/indices/" + itos(k)suffix being at different indices when splitting by/`?

The `/indicies/` properties are created further above, near your first comment. Extra `/` are only possible if someone uses it wrongly.

akien-mga reviewed (commented):
Just left some comments for things which may warrant more eyes on them. I didn't check the existing code to see if what I point out are valid potential problems, I'm just trying to make sure we don't cut corners as PackedScene is very sensitive.

akien-mga commented:
Given the current timetable and the risks from this PR, let's keep it for 4.1 and consider for a 4.0.x cherry-pick once merged and well tested.

In the meantime, it would be good to add a note in the relevant documentation to point to the missing feature.

reduz commented:
I would appreciate the code is commented a bit more because given its hacky, its hard to understand exactly what is going on. Additionally there are some unanswered comments in the review.

YuriSizov commented:
@KoBeWi Bump :)

KoBeWi commented:
> I would appreciate the code is commented a bit more because given its hacky, its hard to understand exactly what is going on.

I checked the code in detail and Node properties were already hacky. This PR just extends the hack to support arrays.

The changes are pretty safe, as they only affect the "deferred NodePath" properties. I rebased/fixed the PR and added some comments.

akien-mga commented:
Thanks!

MaxIsJoe commented:
Does this also fix it for C#?

romifauzi commented:
Tested 4.1 Dev snapshot, upon assigning sprite2d node onto an Array of type Sprite2D, got this error:

```
 Attempted to set a variable of type 'NodePath' into a TypedArray of type 'Object'.
  core/variant/array.cpp:413 - Condition "!_p->typed.validate(value, "set")" is true.
```

Zireael07 commented:
IIRC NodePath is a separate issue

romifauzi commented:
Okay, I was under the assumption this PR fixed issue with dragging nodes into an exported typed array variable @Zireael07. So what does it fixed actually?

KoBeWi commented:
This IS what the PR fixed, but somehow it's broken again on master.

KoBeWi commented:
I opened #76378
It should be included when cherry-picking, because it fixes another issue.

YuriSizov commented:
Skipping this from being cherry-picked for now, as there seem to be problems uncovered by this PR. Once those are addressed, we can cherry-pick in a 4.0.x release. For now, this should be available for testing in 4.1 dev snapshots, if you want to provide feedback and help catch all the bugs.

KoBeWi commented:
The uncovered bugs are corner-case, while this PR fixes a very requested feature. Also the only so far discovered regression is fixed by #76378
Maaybe we can merge it for rc and revert if it causes too many problems? (which i doubt)

YuriSizov commented:
We would prefer to test the changes with 4.1 builds before picking it for a patch release that doesn't get much testing before publishing.

thornySoap commented:
It seems like it's still not working on (at least my) Android export. If I `OS.alert()` the same Array on desktop debug and Android export, on desktop I get an Array with my Nodes, on Android just an empty one.

KoBeWi commented:
@thornySoap Open a new issue, with a minimal reproduction project.
This isn't something that would have platform-specific issues 🤔

YuriSizov commented:
Regarding the cherry-pick, see my comment here: https://github.com/godotengine/godot/pull/76378#issuecomment-1594812425
